### PR TITLE
Support List.NonEmpty representation in Elm

### PR DESCRIPTION
**Pull request by turboMaCk**
This is proof of concept implementation for the https://github.com/Holmusk/elm-street/issues/102.

- In elm `NonEmpty` is represented as `(a, List a)` - this means we don't depend on any type that is not part of `elm/core` but we don't loose any guarantees either.
- In JSON we still use array representation like `[1, 2]`. Because of this NonEmpty can't be defined in terms of `ElmPair` in AST.

## Usage on Elm Side.

One can easily convert pair to any custom implementation of `NonEmpty`:

```elm
type Cons a = Cons a (List a)

uncurry : (a -> b -> c) -> (a, b) -> c
uncurry f (a, b) = f a b

decodeConsInt : Decoder (Cons Int)
decodeConsInt = Decode.map (uncurry Cons) Decoder.generatedDecoder
```

**In most case it should be simpler to use [already compatible implementation](https://package.elm-lang.org/packages/turboMaCk/non-empty-list-alias/latest/)** -- That will especially simplify cases where `NonEmpty` is nested within other data.

### CODE REVIEW

File: frontend/elmapp.config.js [hunk 1]
```diff
@@ -1,10 +1,5 @@
 module.exports = {
     inline: true,
     historyApiFallback: true,
-    proxy:{
-        '/':{
-            target: 'http://localhost:8080',
-            secure: false
-        }
-    }
+    proxy: 'http://localhost:8080'
```
Comment (turboMaCk) [inline frontend/elmapp.config.js:4]: update structure to newer version of create-elm-app (updated in https://github.com/Holmusk/elm-street/pull/103 / https://github.com/Holmusk/elm-street/pull/104)

File: frontend/src/Core/ElmStreet.elm [hunk 2]
```diff
@@ -42,3 +45,8 @@ elmStreetDecodePair decA decB = D.map2 Tuple.pair (D.index 0 decA) (D.index 1 de
 elmStreetDecodeTriple : Decoder a -> Decoder b -> Decoder c -> Decoder (a, b, c)
 elmStreetDecodeTriple decA decB decC = D.map3 (\a b c -> (a,b,c)) (D.index 0 decA) (D.index 1 decB) (D.index 2 decC)
 
+elmStreetDecodeNonEmpty : Decoder a -> Decoder (a, List a)
+elmStreetDecodeNonEmpty decA = D.list decA |> D.andThen (\xs -> case xs of
+                                                 h::t -> D.succeed (h, t)
+                                                 _    -> D.fail "Expecting at least ONE ELEMENT array")
```
Comment (turboMaCk) [inline frontend/src/Core/ElmStreet.elm:51]: we probably should improve formatting of this.

File: src/Elm/Ast.hs [hunk 3]
```diff
@@ -85,6 +85,7 @@ data ElmPrim
     | ElmPair !TypeRef !TypeRef             -- ^ @(A, B)@ in elm
     | ElmTriple !TypeRef !TypeRef !TypeRef  -- ^ @(A, B, C)@ in elm
     | ElmList !TypeRef                      -- ^ @List A@ in elm
+    | ElmNonEmptyPair !TypeRef              -- ^ @NonEmpty A@ represented by @(A, List A)@ in elm
```
Comment (turboMaCk) [inline src/Elm/Ast.hs:88]: We need this extra constructor in AST if we want to represent `NonEmpty` as JSON array. Reusing `ElmPair` would force use to represent `1 :| [ 2, 3]` as `[1, [2,3] ]` in JSON which would be a breaking change for existing APIs.

**Comment by googleson78** [comment]
You could add a reference to your recommended library in the readme, perhaps (?)

**Comment by turboMaCk** [comment]
@googleson78 definitely. Documentation is something I did not look into yet at all but we definitely should review it before considering this or any other implementation complete.

File: src/Elm/Ast.hs (hunk 3 above)
Comment (kwannoel) [reply src/Elm/Ast.hs:88]: How is `ElmPair` represented in JSON?

Comment (turboMaCk) [reply src/Elm/Ast.hs:88]: as an array so `[1]` for singleton or `[1,2,3]` - so flat array.

File: frontend/tests/Tests.elm [hunk 4]
```diff
@@ -41,12 +41,12 @@ defaultOneType =
...
-        , int    = 42
-        , float  = 36.6
-        , text   = "heh"
+        { unit     = ()
+        , bool     = True
+        , char     = 'a'
+        , int      = 42
+        , float    = 36.6
+        , text     = "heh"
         , value  = E.object
```
Comment (turboMaCk) [inline frontend/tests/Tests.elm:50]:
```suggestion
        , value    = E.object
```

File: frontend/tests/Tests.elm [hunk 5]
```diff
@@ -55,12 +55,14 @@ defaultOneType =
...
-        , triple = ('o', False, [0])
-        , list   = [1, 2, 3, 4, 5]
+        , time     = millisToPosix 1550793600000  -- UTCTime (fromGregorian 2019 2 22) 0
+        , maybe    = Just 12
+        , result   = R.Err 666
+        , pair     = ('o', False)
+        , triple   = ('o', False, [0])
+        , list     = [1, 2, 3, 4, 5]
+        , nonEmpty = (1, [])
+
```
Comment (turboMaCk) [inline frontend/tests/Tests.elm:65]:
```suggestion
```

File: types/Types.hs [hunk 6]
```diff
@@ -206,16 +209,17 @@ defaultOneType = OneType
...
-        , primsInt    = 42
-        , primsFloat  = 36.6
-        , primsText   = "heh"
+        { primsUnit     = ()
+        , primsBool     = True
+        , primsChar     = 'a'
+        , primsInt      = 42
+        , primsFloat    = 36.6
+        , primsText     = "heh"
         , primsValue  = object
```
Comment (turboMaCk) [inline types/Types.hs:223]:
```suggestion
        , primsValue   = object
```

File: frontend/src/Core/ElmStreet.elm (hunk 2 above)
Comment (jhrcek) [reply frontend/src/Core/ElmStreet.elm:51]: Feel free to elm-format this (probably in separate PR)

File: src/Elm/Print/Decoder.hs [hunk 7]
```diff
@@ -256,6 +258,15 @@ decodePair = T.unlines
     , "elmStreetDecodePair decA decB = D.map2 Tuple.pair (D.index 0 decA) (D.index 1 decB)"
     ]
 
+-- | @JSON@ decoder Elm help function for List.NonEmpty.
+decodeNonEmpty :: Text
+decodeNonEmpty = T.unlines
+    [ "elmStreetDecodeNonEmpty : Decoder a -> Decoder (a, List a)"
+    , "elmStreetDecodeNonEmpty decA = D.list decA |> D.andThen (\\xs -> case xs of"
+    , "                                                 h::t -> D.succeed (h, t)"
+    , "                                                 _    -> D.fail \"Expecting at least ONE ELEMENT array\")"
```
Comment (jhrcek) [inline src/Elm/Print/Decoder.hs:267]: Perhaps simplify to 
```suggestion
    , "                                                 _    -> D.fail \"Expecting non-empty array\")"
```

**Review by jhrcek — APPROVED** [review]
I like these changes. Just a small suggestion.

File: frontend/src/Core/ElmStreet.elm (hunk 2 above)
Comment (jhrcek) [inline frontend/src/Core/ElmStreet.elm:51]: Also need to update this one
```suggestion
                                                 _    -> D.fail "Expecting non-empty array")
```

**Comment by turboMaCk** [comment]
I'll hold merge of this just for a bit because I also have one other suggestion which is related to this I would like to know your opinion about before merging this PR

**Comment by turboMaCk** [comment]
ok lets go with this first. Other changes will require some discussions as well.

**Pull request merged**